src/lib/ams.c

```diff
@@ -0,0 +1,749 @@
+	dcache_writeback_invalidate_region(ctx->shared->uuid_table,
+					   ctx->shared->uuid_table_size * sizeof(ctx->shared->uuid_table));
+	dcache_writeback_invalidate_region(ctx->shared->routing_table,
+					   ctx->shared->routing_table_size * sizeof(*ctx->shared->routing_table));
+	coherent_release_thread(&ctx->shared->c, sizeof(ctx->shared));
+}
+
+static int am_service_message_send_over_ixc(struct async_message_service *ams, uint32_t slot,
+					    struct ams_routing_entry *target,
+					    void *ctx)
+{
+#if CONFIG_SMP
+	send_message_over_ixc(ams, slot, target);
+#else
+	return -EINVAL;
+#endif
+	return 0;
+}
+
+const struct uuid_idx *am_service_find_uuid_entry_by_uuid(struct am_service_context *const am_service_ctx,
+							  uint8_t const uuid[AM_MESSAGE_UUID_SIZE])
+{
+	uint32_t index;
+
+	if (uuid == NULL)
+		return NULL;
+
+	/* try to find existing entry */
+	for (index = 0; index < am_service_ctx->shared->uuid_table_size; index++) {
+		if (memcmp(am_service_ctx->shared->uuid_table[index].message_uuid,
+			   uuid, AM_MESSAGE_UUID_SIZE) == 0) {
+			return &am_service_ctx->shared->uuid_table[index];
+		}
+	}
+
+	/* and add new one if needed */
+	for (index = 0; index < am_service_ctx->shared->uuid_table_size; index++) {
+		if (am_service_ctx->shared->uuid_table[index].message_type_id == AM_INVALID_MSG_TYPE) {
+			int ec = memcpy_s(am_service_ctx->shared->uuid_table[index].message_uuid,
+					  sizeof(am_service_ctx->shared->uuid_table[index].message_uuid),
+					  uuid, AM_MESSAGE_UUID_SIZE);
+			if (ec != 0)
+				return NULL;
+
+			am_service_ctx->shared->uuid_table[index].message_type_id = ++am_service_ctx->shared->last_used_msg_id;
+			return &am_service_ctx->shared->uuid_table[index];
+		}
+	}
+
+	return NULL;
+}
+
+int am_service_get_message_type_id(const uint8_t message_uuid[AM_MESSAGE_UUID_SIZE],
+				   uint32_t *message_type_id)
+{
+	struct async_message_service *ams = sof_get()->ams;
+	struct uuid_idx const *uuid_entry;
+
+	if (ams->am_service_context == NULL)
+		return -EINVAL;
+
+	*message_type_id = AM_INVALID_MSG_TYPE;
+
+	ams_acquire(ams->am_service_context);
+
+	uuid_entry = am_service_find_uuid_entry_by_uuid(ams->am_service_context, message_uuid);
+	if (uuid_entry == NULL) {
+		ams_release(ams->am_service_context);
+		return -EINVAL;
+	}
+
+	*message_type_id = uuid_entry->message_type_id;
+	ams_release(ams->am_service_context);
+
+	return 0;
+}
+
+int32_t am_service_find_uuid_index_by_msg_type_id(struct am_service_context *const am_service_ctx,
+						  uint32_t const message_type_id)
+{
+	int32_t result = -1;
+	int32_t idx = 0;
+
+	if (message_type_id == AM_INVALID_MSG_TYPE)
+		return -EINVAL;
+
+	for (const struct uuid_idx *iter = am_service_ctx->shared->uuid_table;
+	     iter->message_type_id != AM_INVALID_MSG_TYPE;
+	     iter++) {
+		if (message_type_id == iter->message_type_id) {
+			result = idx;
+			break;
+		}
+		idx++;
+	}
+
+	return result;
```

Comment:
this function seems to be returning non-negatie UUID indices or errno codes (`-EINVAL` above), then this should also be one of those codes, e.g. `return -ENOENT`